lib/charms/observability_libs/v0/juju_topology.py

```diff
@@ -126,29 +126,19 @@ def __init__(
         self._unit = unit
 
     def is_valid_uuid(self, uuid):
-        """Validate the supplied UUID against the Juju Model UUID pattern."""
-        # TODO:
-        # Harness is harcoding an UUID that is v1 not v4: f2c1b2a6-e006-11eb-ba80-0242ac130004
-        # See: https://github.com/canonical/operator/issues/779
-        #
-        # >>> uuid.UUID("f2c1b2a6-e006-11eb-ba80-0242ac130004").version
-        # 1
-        #
-        # we changed the validation of the 3ed UUID block: 4[a-f0-9]{3} -> [a-f0-9]{4}
-        # See: https://github.com/canonical/operator/blob/main/ops/testing.py#L1094
-        #
-        # Juju in fact generates a UUID v4: https://github.com/juju/utils/blob/master/uuid.go#L62
-        # but does not validate it is actually v4:
-        # See:
-        # - https://github.com/juju/utils/blob/master/uuid.go#L22
-        # - https://github.com/juju/schema/blob/master/strings.go#L79
-        #
-        # Once Harness fixes this, we should remove this comment and refactor the regex or
-        # the entire method using the uuid module to validate UUIDs
-        regex = re.compile(
-            "^[a-f0-9]{8}-?[a-f0-9]{4}-?[a-f0-9]{4}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}$"
-        )
-        return bool(regex.match(uuid))
+        """Validate the supplied UUID against the Juju Model UUID pattern.
+
+        Args:
+            uuid: string that needs to be checked if it is valid v4 UUID
+        Returns:
+            if parameter is valid v4 UUID
+        """
+        version = 4
+        try:
+            convert_to_uuid = UUID(uuid, version=version)
+            return str(convert_to_uuid) == uuid
```

Comment:
I would replace these two lines with only this one:

```python
return True if UUID(uuid).version == 4 else False
```

It's simpler, and besides we do not need to convert the uuid and the compare it with the original one.

Reply:
We do need to still check if converting the string to uuid changes the value. The code suggested gives an error on `test_invalid_uuid`. To check that my inquiry was correct, I only altered (from the suggested code) the `return True` to be `return str(UUID(uuid))==uuid` therefore adding the additional step to check if the uuid was being altered. This change passed the `test_invalid_uuid`.
The working function would be:
```
def is_valid_uuid(self, uuid):
        
        try:
            return str(UUID(uuid))==uuid if UUID(uuid).version == 4 else False
        except (ValueError, TypeError):
            return False
```
In this case, since we are repetitively calling `UUID(uuid)`, I feel it would be cleaner to write the code the way that I had done so or, better yet, we can write it even more succinctly into one line as follows:
```
def is_valid_uuid(self, uuid):
        
        try:
            return str(UUID(uuid, version=4)) == uuid
        except (ValueError, TypeError):
            return False
```